### PR TITLE
[firebase-core] add doc comments in source

### PR DESCRIPTION
**Pull request by Simek**
# Why

Fixes ENG-1905

To utilize the auto generation feature for API section on the docs website the package source code must contain the relevant documentation.

# How

This PR moves the current documentation of `FirebaseCore` API - https://docs.expo.io/versions/latest/sdk/firebase-core - with minor tweaks and additions to the package source code.

To full support the migration to automated docs extraction I had to introduce few tweaks to the code:
* ~~`IFirebaseOptions` is now a normal type, `Partial` is assigned in type usages now (the type used inside the Partial need to be defined as separate type to be properly extracted, I have chosen this approach so there will be no new type added)~~
  * after conversation with Hein I have decided to just remove the `Partial` from the type to simplify the things and avoid deprecations or breaking changes.
* `Partial`, `Pick` and `Omit` util types has been linked to the TS handbook

Previously the `IFirebaseOptions` fields were undocumented, I have added some of the descriptions based on Firebase docs, but there are still few missing. Also I'm not sure if all the possible fields are described in this type.

# Test Plan

The changes and generated content have been tests running docs website on `localhost`.

# Preview

Screenshot 2021-09-08 at 12 40 21

# Checklist

- [X] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [X] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [X] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-firebase-core/src/FirebaseCore.ts
Comment (IjzerenHein) [inline packages/expo-firebase-core/src/FirebaseCore.ts:59]: This is a change to the types, we should land this in a separate PR and add a changelog entry for it

File: packages/expo-firebase-core/src/FirebaseOptions.ts
Comment (IjzerenHein) [inline packages/expo-firebase-core/src/FirebaseOptions.ts:3]: Same here

File: docs/components/plugins/__snapshots__/APISection.test.tsx.snap
Comment (IjzerenHein) [inline docs/components/plugins/__snapshots__/APISection.test.tsx.snap:450]: I'd reckon it would be cleaner if these extensions landed in a separate PR

File: docs/components/plugins/api/APISectionUtils.tsx
Comment (IjzerenHein) [inline docs/components/plugins/api/APISectionUtils.tsx:71]: Is `'Partial'` missing here?

File: docs/components/plugins/api/APISectionUtils.tsx
Comment (IjzerenHein) [inline docs/components/plugins/api/APISectionUtils.tsx:88]: 👍

File: packages/expo-firebase-core/src/FirebaseCore.ts
Comment (Simek) [reply packages/expo-firebase-core/src/FirebaseCore.ts:59]: Usually the types changes required for the automatic API generation have landed with the package conversion.

I will update the changelog, however I think that this is not that important change for the users, because as I mentioned in the description I have only moved the `Partial` util placing. Would you like to see additional type instead? 

Also TBH, it's the first usage of `Partial` in Expo types I have seen in last 5 months, and it could be easily removed in favor of optional markings for `IFirebaseOptions`.

File: docs/components/plugins/api/APISectionUtils.tsx
Comment (Simek) [reply docs/components/plugins/api/APISectionUtils.tsx:71]: I wasn't on the non-linkable list, because it's the first use case of `Partial` in the converted packages (~45 packages already converted).

Comment (IjzerenHein) [reply docs/components/plugins/api/APISectionUtils.tsx:71]: Okay, thanks for explaining

File: packages/expo-firebase-core/src/FirebaseCore.ts
Comment (IjzerenHein) [reply packages/expo-firebase-core/src/FirebaseCore.ts:59]: > Usually the types changes required for the automatic API generation have landed with the package conversion.

Okay sounds good, just checking 👍 

> I will update the changelog, however I think that this is not that important change for the users, because as I mentioned in the description I have only moved the Partial util placing. Would you like to see additional type instead, just for the clarity?
>
> Also TBH, it's the first usage of Partial in Expo types I have seen in last 5 months, and it could be easily removed in favor of optional markings for IFirebaseOptions.

Ah I see, I had missed that when reading the PR description, now I see.

As we're changing it from an `interface` to a `type` (which I think is a good thing and it should probably have been a type from the beginning), still keeping the `I` prefix feels wrong here. It will be confusing and I think we need a better migration strategy.

I would suggest to rename `IFirebaseOptions` to `FirebaseOptions` and introduce a new `IFirebaseOptions` type which is marked as deprecated, and which simply refers to `FirebaseOptions`. This way older projects will keep working and we move forward cleaner (imo).

And this should be mentioned in the changelog, as well as `expo-firebase-analytics` and `expo-firebase-recaptcha` should be updated to use the new type (I can help with that).

```
export type FirebaseOptions = {
   ///
};

/**
@deprecated ...
**/
export type IFirebaseOptions = FirebaseOptions;
```

Comment (Simek) [reply packages/expo-firebase-core/src/FirebaseCore.ts:59]: Hmm, there were in the past bigger and breaking changes to the types, but we did not introduce any deprecations so far (just replaced types with not in the changelog):
* https://github.com/expo/expo/pulls?q=is%3Apr+author%3ASimek+doc+comments+is%3Aclosed

It feels like adding deprecation here a bit too much, and an introduction of the additional complexity which we do not need.

There are few things which I want refactor, but so far I was just leaving the tasks on Linear for that, example:
* https://linear.app/expo/issue/ENG-1738/[brightness]-rename-usesystembrightnessasync

Also from the experience from converting other modules I can tell that patterns seen in this package (like `I` in name for the interface), is not propagated across the workspace. It might be something worth discussing, the more consistent approach to naming sounds good, but already converting modules requires a large amount of work, and I just tiring to minimize the changeset size as possible.

**Comment by Simek** [comment]
After latest commit there should be no need to add Changelog entry, since the type changes would not be visible to the package users (it's just different "notation").

File: packages/expo-firebase-core/src/FirebaseCore.ts
Comment (IjzerenHein) [reply packages/expo-firebase-core/src/FirebaseCore.ts]: > Hmm, there were in the past bigger and breaking changes to the types, but we did not introduce any deprecations so far (just replaced types, usually with note in the changelog):
> It feels like adding deprecation here a bit too much, and an introduction of the additional complexity which we do not need.

Okay that sounds reasonable. In that case I would like to opt for removing the `I` and fix it once and for all.

**Review by IjzerenHein — CHANGES_REQUESTED** [review]
The `expo-firebase-recaptcha` build files are outdated, please updating using `yarn build` in `packages/expo-firebase-recaptcha`

File: packages/expo-firebase-core/CHANGELOG.md
Comment (IjzerenHein) [inline packages/expo-firebase-core/CHANGELOG.md:7]:
```suggestion
- Rename `IFirebaseOptions` to `FirebaseOptions` and convert it to a type. ([#14342](https://github.com/expo/expo/pull/14342) by [@Simek](https://github.com/Simek))
```

**Review by IjzerenHein — COMMENTED** [review]
Awesome!

Looks good to me. Because of the breaking change we should probably also change the minimal dependency version on `expo-firebase-core` in package.json.
I'm not entirely sure what is good practise there. To do this now have this done when releasing SDK 43. @tsapeta what are your thoughts?

**Review by IjzerenHein — APPROVED** [review]

File: packages/expo-firebase-core/CHANGELOG.md
Comment (Simek) [reply packages/expo-firebase-core/CHANGELOG.md:7]: hmm, it was a `type` before, the only difference is that is no longer wrapped with `Partial`:
* https://github.com/expo/expo/pull/14342/files#diff-3d6e54fb5b9a09bb5225cb97b40ed3ec93d8abbf9013eb383b3e780bfc43becdL3

Comment (IjzerenHein) [reply packages/expo-firebase-core/CHANGELOG.md:7]: ow, huh, I wasn't aware of that
now things are starting to make sense..
well in that case, yeah please ignore that suggestion

**Comment by Simek** [comment]
@IjzerenHein Thank you for the review and suggestion on this PR! 🙏 

I have updated the changelog anyway, I don't think that the reason of rename was that important anyway. 🙂 

However, I would like to wait for the @tsapeta approval before the merge, since you have brought up the topic of handling this breaking change.

**Review by bbarthec — APPROVED** [review]
It looks good to me 👍

**Review by tsapeta — APPROVED** [review]

**Pull request merged**
